images/encryption/blockcipher/blockcipher_siv.go

```diff
@@ -0,0 +1,219 @@
+	CryptedDataReader
+}
+
+type aessivcryptor struct {
+	bc           *AESSIVLayerBlockCipher
+	outputReader io.Reader
+	digester     digest.Digester
+}
+
+// NewAESSIVLayerBlockCipher returns a new AES SIV block cipher of 256 or 512 bits
+func NewAESSIVLayerBlockCipher(bits int) (LayerBlockCipher, error) {
+	if bits != 256 && bits != 512 {
+		return nil, errors.New("AES SIV bit count not supported")
+	}
+	return &AESSIVLayerBlockCipher{bits: bits}, nil
+}
+
+func (r *aessivcryptor) Read(p []byte) (int, error) {
+	if r.bc.err != nil {
+		return 0, r.bc.err
+	}
+
+	for {
+		// return data if we have any
+		if r.bc.outbuffer != nil && r.bc.outoffset < len(r.bc.outbuffer) {
+			n := copy(p, r.bc.outbuffer[r.bc.outoffset:])
+			r.bc.outoffset += n
+			r.digester.Hash().Write(p[:n])
+
+			return n, nil
+		}
+		// no data and hit eof before?
+		if r.bc.eof {
+			return 0, io.EOF
+		}
+		// read new data; we expect to get r.bc.toread number of bytes
+		// for anything less we assume it's EOF
+		numbytes := 0
+		for numbytes < r.bc.toread {
+			var n int
+			n, r.bc.err = r.bc.reader.ReadAt(r.bc.inbuffer[numbytes:r.bc.toread-numbytes], r.bc.inoffset+int64(numbytes))
+			numbytes += n
+			if r.bc.err != nil {
+				if r.bc.err == io.EOF {
+					r.bc.eof = true
+					r.bc.err = nil
+					break
+				} else {
+					return 0, r.bc.err
+				}
+			}
+			if n == 0 {
+				break
+			}
+		}
+		if numbytes < r.bc.toread {
+			r.bc.eof = true
+		}
+
+		r.bc.inoffset += int64(numbytes)
+
+		// transform the data
+		if r.bc.encryptor != nil {
+			r.bc.outbuffer = r.bc.encryptor.Seal(nil, r.bc.inbuffer[:numbytes], []byte(""), r.bc.eof)
+		} else {
+			r.bc.outbuffer, r.bc.err = r.bc.decryptor.Open(nil, r.bc.inbuffer[:numbytes], []byte(""), r.bc.eof)
+			if r.bc.err != nil {
+				return 0, r.bc.err
+			}
+		}
+		// let reader start from beginning of buffer
+		r.bc.outoffset = 0
+		r.bc.outsize += int64(len(r.bc.outbuffer))
+	}
+}
+
+// Size returns the Size of the encrypted data
+func (r *aessivcryptor) Size() int64 {
+	return r.bc.outsize
+}
+
+func (r *aessivcryptor) Digest() digest.Digest {
+	return r.digester.Digest()
+}
+
+func (r *aessivcryptor) Hash() hash.Hash {
+	return r.digester.Hash()
+}
+
+// init initializes an instance
+func (bc *AESSIVLayerBlockCipher) init(encrypt bool, reader io.ReaderAt, opt LayerBlockCipherOptions) (LayerBlockCipherOptions, error) {
+	var (
+		err   error
+		nonce []byte
+		se    miscreant.StreamEncryptor
+	)
+
+	bc.reader = reader
+
+	key := opt.SymmetricKey
+	if len(key) != bc.bits/8 {
+		return LayerBlockCipherOptions{}, fmt.Errorf("invalid key length of %d bytes; need %d bytes", len(key), bc.bits/8)
+	}
+
+	nonceStr := opt.CipherOptions["nonce"]
+	if nonceStr != "" {
+		nonce, err = base64.StdEncoding.DecodeString(nonceStr)
+		if err != nil {
+			return LayerBlockCipherOptions{}, errors.New("Failed to decode nonce")
+		}
+	} else {
+		nonce = make([]byte, se.NonceSize())
+		if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
```

Comment:
Miscreatnt provides a `GenerateNonce` function https://godoc.org/github.com/miscreant/miscreant-go#GenerateNonce

Reply:
Hm, we are using NewStreamDecryptor which returns us
```
type StreamDecryptor struct {
	// cipher.AEAD instance underlying this STREAM
	a cipher.AEAD

	// Nonce encoder instance which computes per-message nonces
	n *nonceEncoder32
}
```

https://github.com/miscreant/miscreant.go/blob/master/stream.go#L77

I am missing a function that gives me access to 'a'.

Reply:
... might as well raise an issue with the miscreant project

Reply:
https://github.com/miscreant/miscreant.go/issues/4

Reply:
Ah sorry I was misreading the API docs.

Reply:
So we leave it as it is now ?